### PR TITLE
Nvshmem refit

### PR DESCRIPTION
**Pull request by wdykas**
# What does this PR do ?
The PR adds the Nvshmem backend from @kanz-nv to the Megatron RL loop.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/resharding/copy_services/gloo_copy_service.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    """
+
+    def __init__(self):
+        self.rank = dist.get_rank()
+        self.world_size = dist.get_world_size()
+        self.gloo_pg = dist.new_group(backend="gloo")
+        self.send_ops: List[SendOp] = []
+        self.recv_ops: List[Tuple[RecvOp, torch.Tensor]] = []
+        self._copy_stream = torch.cuda.Stream()
+        logger.info(f"GlooCopyService initialized on rank {self.rank} with {self.world_size} ranks")
```
Comment (Skylion007) [inline megatron/core/resharding/copy_services/gloo_copy_service.py]: slight nit, use lazy string implicit % formatting for logger here.

File: megatron/core/resharding/execution.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    # Execute
+    logger.info(f"Executing {len(plan.send_ops)} sends + {len(plan.recv_ops)} recvs")
+    service.run()
+    # TODO(Peter) remove this eventually?
+    dist.barrier()
+
+    # Write back received buffers into their destination parameter slices
+    for recv_buffer, dst_param, dst_slice in recv_writebacks:
+        with torch.no_grad():
```
Comment (Skylion007) [inline megatron/core/resharding/execution.py:63]: Slight nit, put for loop inside of with context. No need to reneter and exit with context repeatdely.

File: megatron/core/resharding/nvshmem_copy_service/core/gpu_resource_manager.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        # Initialize NVSHMEM with the broadcasted UID
+        nvshmem.core.init(
+            device=self.device,
+            uid=broadcast_objects[0],
+            rank=rank_id,
+            nranks=num_ranks,
+            initializer_method="uid",
+        )
+
+        print("NVSHMEM initialized")
```
Comment (Skylion007) [inline megatron/core/resharding/nvshmem_copy_service/core/gpu_resource_manager.py]: Should these print statements be logger calls?

File: megatron/core/resharding/refit.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    src_lm = src_model[0] if isinstance(src_model, (list, tuple)) else src_model
+    tgt_lm = target_model[0] if isinstance(target_model, (list, tuple)) else target_model
+
+    num_experts = src_lm.config.num_moe_experts
+
+    # Unwrap to get owning modules (with parameters and pg_collection)
+    src_core = unwrap_model(src_lm)
+    tgt_core = unwrap_model(tgt_lm)
+
+    # Ensure pg_collection exists
```
Comment (kanz-nv) [inline megatron/core/resharding/refit.py]: what is pg_collection?

**Comment by wdykas** [comment]
/ok to test [a4e7d4f](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/a4e7d4f02a227974c909964ec4abc75e0daea068)

**Comment by wdykas** [comment]
/ok to test [23ec73d](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/23ec73dbd33d107e0a66a65d28c9b551fce4597c)

**Comment by wdykas** [comment]
/ok to test [34990da](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/34990da9de91be0956c967505aeeecf7ceba3d44)

**Comment by wdykas** [comment]
/ok to test [ae71cec](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/ae71cecca2620ee0b95d36bfa337ac200ab962a1)

**Comment by wdykas** [comment]
/ok to test [3bf9528](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/3bf9528c2315f3fc206ce95abd1073a805833334)

**Comment by wdykas** [comment]
/ok to test [3817e28](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/3817e28dcd2b92a17559c21cad668db8a2310989)

**Comment by wdykas** [comment]
/ok to test [5c1c58f](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/5c1c58fe0fe4271de3e68c477f9f7e1644066810)

**Comment by wdykas** [comment]
/ok to test [88959e8](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/88959e8c5a3360bd4557a2f64d36db561f9d9c27)

**Comment by wdykas** [comment]
/ok to test [4306ccd](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/4306ccd18672434851ce83d1d31be21193d0ad58)

**Comment by wdykas** [comment]
/ok to test [53091cb](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/53091cb11629f1a169754437382cebbe0d8b7090)

**Comment by wdykas** [comment]
@santhnm2 @kanz-nv @sidsingh-nvidia @kvareddy Here is the MR now for Kan's Nvshmem backend. Please take a look when you get the chance. The follow up MRs after this are wiring through PP support which I have a branch for then wiring through the MoE EP refit through RL. The functionality for those for refit is already in our backend it just making sure it works properly with RL

**Comment by wdykas** [comment]
/ok to test [e0b3fc4](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/e0b3fc4da3c01a6b2ae86d691fe1bb0e6d19a1c7)

**Comment by wdykas** [comment]
/ok to test [1190730](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/1190730dc79cf94c621e10ea99a62825afd82897)

File: megatron/core/resharding/nvshmem_copy_service/core/pipeline_executor.py [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+                )
+                torch.cuda.nvtx.range_pop()
+
+            # Ensure send completes
+            self.send_stream.sync()
+            nvshmem.core.quiet(stream=self.send_stream)
+
+            # Step 4: Global barrier
+            torch.cuda.nvtx.range_push("Step 4: Barrier")
+            nvshmem.core.barrier_all(stream=self.send_stream)
```
Comment (sidsingh-nvidia) [inline megatron/core/resharding/nvshmem_copy_service/core/pipeline_executor.py:181]: Have we confirmed that this is not a performance bottleneck?

Comment (wdykas) [reply megatron/core/resharding/nvshmem_copy_service/core/pipeline_executor.py:181]: I am not sure

**Comment by wdykas** [comment]
/ok to test [284751a](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/284751a72250397e6bfa8dfa0a0bee94a5c76e5f)

File: megatron/core/resharding/nvshmem_copy_service/core/pipeline_executor.py (hunk 5 above)
Comment (wdykas) [reply megatron/core/resharding/nvshmem_copy_service/core/pipeline_executor.py:181]: @sidsingh-nvidia does this barrier the streams every iteration?

Comment (wdykas) [reply megatron/core/resharding/nvshmem_copy_service/core/pipeline_executor.py:181]: verified this can be a slow down if the schedule is not setup well but assuming the schedule is fine we should be ok.

**Review by santhnm2 — APPROVED** [review]

**Review by sidsingh-nvidia — APPROVED** [review]

**Review by kvareddy — APPROVED** [review]

**Comment by wdykas** [comment]
/ok to test [d78b710](https://github.com/NVIDIA/Megatron-LM/pull/2696/commits/d78b710e8d468f93657924be9414d2e8e51363dd)
